File: go/spore-core/harness_catalogue_test.go
package sporecore

import (
	"context"
	"encoding/json"
	"testing"
)

// Issue #81 (Q4b): a tool returning AwaitingClarification pauses the loop with a
// PausedState whose HumanRequest is a Clarification (NO child_state), and the
// clarifying call is preserved as the HEAD of PendingToolCalls.
func TestAwaitingClarificationPausesWithClarification(t *testing.T) {
	a := NewMockAgent("t")
	a.Push(NewToolCallRequested([]ToolCall{
		{ID: "ask1", Name: "ask_user_question", Input: json.RawMessage(`{"question":"which?","options":["a","b"]}`)},
	}, turnUsage()))
	cfg := standardCfg(a)
	reg := NewScriptedToolRegistry()
	opts := []string{"a", "b"}
	reg.Push(ToolOutput{Kind: ToolOutputAwaitingClarification, Question: "which?", Options: &opts})
	cfg.ToolRegistry = reg
	h := NewStandardHarness(cfg)
	r := h.Run(context.Background(), NewHarnessRunOptions(reactTask(5)))

	if r.Kind != RunWaitingForHuman {
		t.Fatalf("expected WaitingForHuman, got %q", r.Kind)
	}
	if r.Request == nil || r.Request.Kind != HumanReqClarification || r.Request.Question != "which?" {
		t.Fatalf("request: %+v", r.Request)
	}
	if r.Request.Options == nil || len(*r.Request.Options) != 2 {
		t.Fatalf("request options: %+v", r.Request.Options)
	}
	if r.State == nil {
		t.Fatalf("expected paused state")
	}
	if r.State.ChildState != nil {
		t.Fatalf("clarification pause must have NO child_state")
	}
	if len(r.State.PendingToolCalls) != 1 || r.State.PendingToolCalls[0].ID != "ask1" {
		t.Fatalf("clarifying call must be head of pending: %+v", r.State.PendingToolCalls)
	}
}

// Issue #81 (Q4b): resuming a Clarification pause injects the human's answer as
// the tool RESULT for the clarifying call, then continues the loop.
func TestResumeClarificationInjectsAnswerAsToolResult(t *testing.T) {
	a := NewMockAgent("t")
	// After resume the loop runs the agent again; it produces a final response.
	a.Push(NewFinalResponse("done-after-answer", turnUsage()))
	cfg := standardCfg(a)
	cm := &recordingContextManager{}
	cfg.ContextManager = cm
	h := NewStandardHarness(cfg)

	req := HumanRequest{Kind: HumanReqClarification, Question: "which?"}
	state := PausedState{
		SessionID:        SessionID("s1"),
		TaskID:           reactTask(5).ID,
		TurnNumber:       1,
		SessionState:     SessionState{},
		PendingToolCalls: []ToolCall{{ID: "ask1", Name: "ask_user_question", Input: json.RawMessage(`{"question":"which?"}`)}},
		HumanRequest:     &req,
		Task:             reactTask(5),
	}
	r := h.Resume(context.Background(), state, HumanResponse{Kind: HumanRespAnswer, Text: "option-a"}, nil)
	if r.Kind != RunSuccess || r.Output != "done-after-answer" {
		t.Fatalf("expected success after answer, got %+v", r)
	}
	// The answer must have been injected as the tool RESULT for the clarifying
	// call ("ask1") — a Tool-role row keyed by that call id — NOT appended as a
	// free-standing user message.
	var injectedAsToolResult bool
	for i, role := range cm.roles {
		if role == RoleTool && cm.callIDs[i] == "ask1" {
			injectedAsToolResult = true
		}
		if role == RoleUser && cm.texts[i] == "option-a" {
			t.Fatalf("answer must NOT be appended as a free-standing user message")
		}
	}
	if !injectedAsToolResult {
		t.Fatalf("answer not injected as a tool result for the clarifying call: roles=%+v", cm.roles)
	}
}

// Issue #81: the send_message tool surfaces a HarnessStreamEvent of kind
// user_message and records a minimal success result so the loop continues.
func TestSendMessageEmitsUserMessageEvent(t *testing.T) {
	a := NewMockAgent("t")
	a.Push(NewToolCallRequested([]ToolCall{
		{ID: "m1", Name: "send_message", Input: json.RawMessage(`{"content":"hello human"}`)},
	}, turnUsage()))
	a.Push(NewFinalResponse("done", turnUsage()))
	cfg := standardCfg(a)
	reg := NewScriptedToolRegistry()
	reg.Push(ToolOutput{Kind: ToolOutputSuccess, Content: "hello human"})
	cfg.ToolRegistry = reg
	h := NewStandardHarness(cfg)

	var events []HarnessStreamEvent
	sink := func(e HarnessStreamEvent) { events = append(events, e) }
	opts := NewHarnessRunOptions(reactTask(5))
	opts.OnStream = sink
	r := h.Run(context.Background(), opts)
	if r.Kind != RunSuccess {
		t.Fatalf("expected success, got %+v", r)
	}
	var found bool
	for _, e := range events {
		if e.Kind == HarnessStreamUserMessage {
			found = true
			if e.Content != "hello human" {
				t.Fatalf("user_message content: %q", e.Content)
			}
		}
	}
	if !found {
		t.Fatalf("expected a user_message stream event")
	}
}

// recordingContextManager is defined in harness_test.go (same package); reused
// here to assert the clarification-resume injection path.

File: typescript/packages/core/tests/harness-catalogue.test.ts
/**
 * Harness-loop tests for the #81 Standard Tool Catalogue wiring:
 *  - `send_message` → the loop emits a `user_message` stream event and records a
 *    minimal success result (it does NOT collapse to a normal tool result path).
 *  - `awaiting_clarification` → the loop pauses into a `PausedState` directly
 *    (NO ChildPausedState), sets `human_request` to `clarification`, preserves
 *    the clarifying call as head of `pending_tool_calls`, and returns
 *    `waiting_for_human`. On resume the answer text is injected as that call's
 *    tool result and the loop continues.
 *
 * Mirrors the corresponding Rust harness tests for issue #81.
 */

import { describe, expect, it } from "vitest";

import {
  AgentId,
  MockAgent,
  SessionId,
  StandardHarness,
  newTask,
  type HarnessConfig,
  type HarnessStreamEvent,
  type LoopStrategy,
  type Task,
  type TokenUsage,
  type ToolCall,
  type ToolOutput,
  type TurnResult,
} from "../src/index.js";

import {
  AllowAllSandbox,
  AlwaysContinuePolicy,
  NoopContextManager,
  ScriptedToolRegistry,
  registryWith,
} from "../src/harness/testing.js";

function standardConfig(agent: MockAgent): HarnessConfig {
  return {
    registry: registryWith({ agent }),
    toolRegistry: new ScriptedToolRegistry(),
    sandbox: new AllowAllSandbox(),
    contextManager: new NoopContextManager(),
    terminationPolicy: new AlwaysContinuePolicy(),
    modelParams: { stop_sequences: [] },
  };
}

function react(max: number): Task {
  const strategy: LoopStrategy = {
    kind: "react",
    budget: { kind: "per_loop", value: max },
    agent: "",
    toolset: "",
  };
  return newTask("do something", SessionId.of("s1"), strategy);
}

function usage(): TokenUsage {
  return { input_tokens: 1, output_tokens: 1, cache_read_tokens: null, cache_write_tokens: null };
}

function agentCalling(name: string, input: unknown, callId = "c0"): MockAgent {
  const a = new MockAgent(AgentId.of("test"));
  a.push({
    kind: "tool_call_requested",
    calls: [{ id: callId, name, input } as ToolCall],
    usage: usage(),
  } as TurnResult);
  a.push({ kind: "final_response", content: "done", usage: usage() } as TurnResult);
  return a;
}

// ============================================================================
// send_message → user_message stream event
// ============================================================================

describe("send_message loop wiring (#81)", () => {
  it("emits a user_message stream event and records a success result", async () => {
    const a = agentCalling("send_message", { content: "hello human" });
    const cfg = standardConfig(a);
    cfg.toolRegistry = new ScriptedToolRegistry().push({
      kind: "success",
      content: "hello human",
      truncated: false,
    } as ToolOutput);
    const h = new StandardHarness(cfg);

    const events: HarnessStreamEvent[] = [];
    const r = await h.run({ task: react(5), on_stream: (e) => events.push(e) });

    expect(r.kind).toBe("success");
    const userMessages = events.filter((e) => e.kind === "user_message");
    expect(userMessages.length).toBe(1);
    expect(userMessages[0]).toEqual({ kind: "user_message", content: "hello human" });
  });

  it("a send_message that errors emits no user_message event", async () => {
    const a = agentCalling("send_message", {});
    const cfg = standardConfig(a);
    cfg.toolRegistry = new ScriptedToolRegistry().push({
      kind: "error",
      message: "invalid parameters: content required",
      recoverable: true,
    } as ToolOutput);
    const h = new StandardHarness(cfg);

    const events: HarnessStreamEvent[] = [];
    await h.run({ task: react(5), on_stream: (e) => events.push(e) });
    expect(events.filter((e) => e.kind === "user_message").length).toBe(0);
  });
});

// ============================================================================
// awaiting_clarification → pause + resume
// ============================================================================

describe("awaiting_clarification loop wiring (#81)", () => {
  it("pauses into a PausedState (no child_state) with a clarification request", async () => {
    const a = agentCalling("ask_user_question", { question: "which?", options: ["a", "b"] });
    const cfg = standardConfig(a);
    cfg.toolRegistry = new ScriptedToolRegistry().push({
      kind: "awaiting_clarification",
      question: "which?",
      options: ["a", "b"],
    } as ToolOutput);
    const h = new StandardHarness(cfg);

    const r = await h.run({ task: react(5) });
    expect(r.kind).toBe("waiting_for_human");
    if (r.kind !== "waiting_for_human") return;

    // No ChildPausedState (this is NOT the subagent path).
    expect(r.state.child_state).toBeNull();
    // human_request is a clarification carrying the options.
    expect(r.request.kind).toBe("clarification");
    if (r.request.kind === "clarification") {
      expect(r.request.question).toBe("which?");
      expect(r.request.options).toEqual(["a", "b"]);
    }
    // The clarifying call is preserved as the HEAD of pending_tool_calls.
    expect(r.state.pending_tool_calls.length).toBe(1);
    expect(r.state.pending_tool_calls[0]?.name).toBe("ask_user_question");
    expect(r.state.pending_tool_calls[0]?.id).toBe("c0");
  });

  it("resume injects the answer as the clarifying call's tool result and continues", async () => {
    const a = agentCalling("ask_user_question", { question: "which?" });
    const cfg = standardConfig(a);
    cfg.toolRegistry = new ScriptedToolRegistry().push({
      kind: "awaiting_clarification",
      question: "which?",
    } as ToolOutput);
    const h = new StandardHarness(cfg);

    const paused = await h.run({ task: react(5) });
    expect(paused.kind).toBe("waiting_for_human");
    if (paused.kind !== "waiting_for_human") return;

    const resumed = await h.resume(paused.state, { kind: "answer", text: "use a" });
    expect(resumed.kind).toBe("success");
    if (resumed.kind !== "success") return;
    expect(resumed.output).toBe("done");

    // The clarifying call's tool result is the injected answer text — recorded
    // as a tool-role message, NOT a free-standing user message.
    const toolMsgs = paused.state.session_state.messages.filter((m) => m.role === "tool");
    expect(toolMsgs.length).toBeGreaterThanOrEqual(1);
    const hasAnswer = paused.state.session_state.messages.some(
      (m) => m.role === "tool" && m.content.type === "text" && m.content.text === "use a",
    );
    expect(hasAnswer).toBe(true);
  });

  it("a clarification pause with options round-trips through HumanRequest back-compat", async () => {
    // Back-compat: a clarification HumanRequest without `options` still
    // deserializes (the field is optional).
    const withOptions = { kind: "clarification", question: "q", options: ["x"] };
    const back = JSON.parse(JSON.stringify(withOptions));
    expect(back.options).toEqual(["x"]);
    const bare = JSON.parse(JSON.stringify({ kind: "clarification", question: "q" }));
    expect(bare.options).toBeUndefined();
  });
});
